packages/backend-modules/publikator/script/migrateAudioSources.ts

```diff
@@ -0,0 +1,200 @@
+
+const author = { name: 'Publikator Bot', email: 'tech@republik.ch' }
+
+const handleBatch = async (rows: any[], count: number, pgdb: any) => {
+  debug('handleBatch begin. rows: %i', rows?.length)
+
+  await Promise.mapSeries(rows, async (row) => {
+    const { repoId, archivedAt } = row
+    debug(
+      'repoId: %s (%s)',
+      repoId,
+      `https://publikator.republik.ch/repo/${repoId}/tree`,
+    )
+
+    if (archivedAt) {
+      debug('repo archived. done')
+      console.log(
+        'repo archived',
+        repoId,
+        `https://publikator.republik.ch/repo/${repoId}/tree`,
+      )
+      return
+    }
+
+    const publication = await pgdb.publikator.milestones.findOne({
+      repoId,
+      scope: 'publication',
+      'publishedAt !=': null,
+      revokedAt: null,
+    })
+    debug('publication: %s', publication?.id)
+
+    const isPublished = !!publication
+    debug('isPublished: %o', isPublished)
+
+    const publicationCommit =
+      isPublished &&
+      (await pgdb.publikator.commits.findOne({ id: publication.commitId }))
+    debug('publicationCommit: %s', publicationCommit?.id)
+
+    const latestCommit = await pgdb.publikator.commits.findOne(
+      { repoId },
+      { orderBy: { createdAt: 'DESC' }, limit: 1 },
+    )
+    debug('latestCommit: %s', latestCommit?.id)
+
+    const hasDuration = !!latestCommit?.meta?.audioSourceDurationMs
+    debug('hasDuration: %o', hasDuration)
+    if (hasDuration) {
+      debug('meta.hasDuration present. done')
+      console.log(
+        'meta.hasDuration present',
+        repoId,
+        `https://publikator.republik.ch/repo/${repoId}/tree`,
+      )
+      return
+    }
+
+    const workingCommit = publicationCommit || latestCommit
+    debug('workingCommit: %s', workingCommit?.id)
+
+    const { meta, type, content, content__markdown } = workingCommit
+
+    const hasAudioSourceMp3 = !!meta.audioSourceMp3
+    if (!hasAudioSourceMp3) {
+      debug('meta.hasAudioSourceMp3 not present. done')
+      console.log(
+        'meta.hasAudioSourceMp3 not present',
+        repoId,
+        `https://publikator.republik.ch/repo/${repoId}/tree`,
+      )
+      return
+    }
+
+    await maybeApplyAudioSourceDuration(meta)
+
+    const tx = await pgdb.transactionBegin()
+    debug('transaction begin')
+
+    try {
+      const commit = await tx.publikator.commits.insertAndGet({
+        repoId,
+        content: content || mdastParse(content__markdown),
+        type: type || 'mdast',
+        meta,
+        message: 'Audio-Datei vermessen',
+        author,
+        ...(workingCommit && { parentIds: [workingCommit.id] }),
+        createdAt: moment(workingCommit.createdAt).add(1, 'second'),
+      })
+      debug('inserted commit: %s', commit.id)
+      console.log(
+        'inserted commit',
+        repoId,
+        `https://publikator.republik.ch/repo/${repoId}/tree`,
+      )
+
+      if (isPublished) {
```

Comment:
Why do you need to replace the derivative, and can't just update the commitId?

Reply:
Technically, we could just update the commit ID. But that would lead to some unwanted side-effects.

A derivative is "derived" from a commit. Could be a PDF version of a commit, or a synthetic read aloud audio file. An information or assets which belongs to a commit, but is appended asynchronously.

Once you publish a commit with derivates, they become available, too. (In fact, synthetic read aloud audio files are provided that way.)

If you were to add a commit, update its dervatives to that new commit and publish it it, it be working fine, too.

Now, if you or someone else decides to publish the previous commit again (aka. reverting to a previous version) its derivates would have been gone.

To prevent this, derivatives are copied if a commit was previously published.